File: app/ios/OffTheCloud/OffTheCloud/ConnectionEndpointFields.swift
// SPDX-License-Identifier: AGPL-3.0-or-later

//
//  ConnectionEndpointFields.swift
//  OffTheCloud
//
//  Issue #121: how a device's address is entered, shared by the first-run
//  screen and Settings so the two can't drift.
//
//  A device on the public bridge is reachable at wss://<name>.off-the.cloud
//  /ws, and nothing about that is the owner's to decide but the name - so
//  the name is all the form asks for. A device reached any other way (own
//  bridge, Tailscale Funnel, the LAN) needs its address typed, behind a
//  toggle, and an existing custom address opens the form in that mode so
//  it never silently rewrites what someone set up on purpose.
//

import SwiftUI

struct ConnectionEndpointFields: View {
    /// The endpoint as stored - always the full address, whichever way it
    /// was entered.
    @Binding var endpoint: String

    @State private var deviceName = ""
    @State private var customAddress = false
    @State private var loaded = false

    var body: some View {
        Group {
            if customAddress {
                TextField("Address (wss://host/ws)", text: $endpoint)
                    .autocapitalization(.none)
                    .disableAutocorrection(true)
                    .textContentType(.URL)
                    .keyboardType(.URL)
            } else {
                HStack(spacing: 0) {
                    TextField("Device name", text: $deviceName)
                        .autocapitalization(.none)
                        .disableAutocorrection(true)
                        .keyboardType(.asciiCapable)
                        .onChange(of: deviceName) { _, name in
                            endpoint = name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
                                ? "" : SecretsStore.bridgeEndpoint(forName: name)
                        }
                    Text(".\(SecretsStore.bridgeDomain)")
                        .foregroundStyle(.secondary)
                        .lineLimit(1)
                        .fixedSize()
                }
            }
            Toggle("Custom address", isOn: $customAddress)
                .onChange(of: customAddress) { _, custom in
                    // Switching to the name keeps whatever name the
                    // address had; switching to custom keeps the address.
                    if !custom {
                        deviceName = SecretsStore.bridgeName(fromEndpoint: endpoint) ?? ""
                        endpoint = deviceName.isEmpty ? "" : SecretsStore.bridgeEndpoint(forName: deviceName)
                    }
                }
            if customAddress {
                Text("For a device not on the Off The Cloud bridge - your own bridge, Tailscale Funnel, or the local network (ws://192.168.…:8080/ws).")
                    .font(.caption).foregroundStyle(.secondary)
            }
        }
        .onAppear {
            guard !loaded else { return }
            loaded = true
            if let name = SecretsStore.bridgeName(fromEndpoint: endpoint) {
                deviceName = name
            } else if !endpoint.isEmpty {
                customAddress = true
            }
        }
    }
}

File: app/android/app/src/main/java/cloud/offthe/otc/ui/common/ConnectionEndpointFields.kt
// SPDX-License-Identifier: AGPL-3.0-or-later
package cloud.offthe.otc.ui.common

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.MaterialTheme
import cloud.offthe.otc.ui.common.OTCTextField
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.input.KeyboardCapitalization
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import cloud.offthe.otc.data.SecretsStore

// Port of ConnectionEndpointFields.swift (issue #121): the device's name
// is all the bridge needs; a custom address is behind a toggle, and an
// existing custom address opens the form in that mode.
@Composable
fun ConnectionEndpointFields(endpoint: String, onEndpointChange: (String) -> Unit) {
    var deviceName by remember { mutableStateOf("") }
    var customAddress by remember { mutableStateOf(false) }
    var loaded by remember { mutableStateOf(false) }

    LaunchedEffect(Unit) {
        if (loaded) return@LaunchedEffect
        loaded = true
        val name = SecretsStore.bridgeName(endpoint)
        if (name != null) deviceName = name else if (endpoint.isNotEmpty()) customAddress = true
    }

    Column {
        if (customAddress) {
            OTCTextField(
                value = endpoint, onValueChange = onEndpointChange,
                label = { Text("Address (wss://host/ws)") }, singleLine = true,
                keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Uri, capitalization = KeyboardCapitalization.None),
                modifier = Modifier.fillMaxWidth(),
            )
        } else {
            Row(verticalAlignment = Alignment.CenterVertically) {
                OTCTextField(
                    value = deviceName,
                    onValueChange = { name ->
                        deviceName = name
                        onEndpointChange(if (name.trim().isEmpty()) "" else SecretsStore.bridgeEndpoint(name))
                    },
                    label = { Text("Device name") }, singleLine = true,
                    keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Ascii, capitalization = KeyboardCapitalization.None),
                    modifier = Modifier.weight(1f),
                )
                Text(".${SecretsStore.bridgeDomain}", color = MaterialTheme.colorScheme.onSurfaceVariant, modifier = Modifier.padding(start = 6.dp))
            }
        }
        Row(verticalAlignment = Alignment.CenterVertically, modifier = Modifier.fillMaxWidth().padding(top = 8.dp)) {
            Text("Custom address", modifier = Modifier.weight(1f))
            Switch(checked = customAddress, onCheckedChange = { custom ->
                customAddress = custom
                if (!custom) {
                    deviceName = SecretsStore.bridgeName(endpoint) ?: ""
                    onEndpointChange(if (deviceName.isEmpty()) "" else SecretsStore.bridgeEndpoint(deviceName))
                }
            })
        }
        if (customAddress) {
            Text(
                "For a device not on the Off The Cloud bridge - your own bridge, Tailscale Funnel, or the local network (ws://192.168.…:8080/ws).",
                style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
    }
}
